lane.py

```diff
@@ -387,897 +372,812 @@ def has_next_page(self):
-          `self.include_subgenres` is True, any of those genres'
-          subgenres).
+    # This has no effect unless list_datasource_id or
+    # list_identifier_id is also set. If this is set, then a book will
+    # only be shown if it has a CustomListEntry on an appropriate list
+    # where `most_recent_appearance` is within this number of days. If
+    # the number is zero, then the lane contains _every_ book with a
+    # CustomListEntry associated with an appropriate list.
+    list_seen_in_previous_days = Column(Integer, nullable=True)
+
+    # If this is set to True, then a book will show up in a lane only
+    # if it would _also_ show up in its parent lane.
+    inherit_parent_restrictions = Column(Boolean, default=False, nullable=False)
+
+    # Patrons whose external type is in this list will be sent to this
+    # lane when they ask for the root lane.
+    #
+    # This is almost never necessary.
+    root_for_patron_type = Column(ARRAY(Unicode), nullable=True)
+
+    # Only a visible lane will show up in the user interface.  The
+    # admin interface can see all the lanes, visible or not.
+    _visible = Column(Boolean, default=True, nullable=False, name="visible")
+
+    __table_args__ = (
+        UniqueConstraint('library_id', 'identifier'),
+        UniqueConstraint('parent_id', 'display_name'),
+    )
 
-        * Have the same appeal as `self.appeal`, if `self.appeal` is present.
+    def get_library(self, _db):
+        """For compatibility with WorkList.get_library()."""
+        return self.library
 
-        * Are intended for the audience in `self.audience`.
+    @property
+    def collection_ids(self):
+        return [x.id for x in self.library.collections]
 
-        * Are fiction (if `self.fiction` is True), or nonfiction (if fiction
-          is false), or of the default fiction status for the genre
-          (if fiction==FICTION_DEFAULT_FOR_GENRE and all genres have
-          the same default fiction status). If fiction==None, no fiction
-          restriction is applied.
+    @property
+    def visible_children(self):
+        children = [lane for lane in self.sublanes if lane.visible]
+        return sorted(children, key=lambda x: x.display_name)
 
-        * Have a delivery mechanism that can be rendered by the
-          default client.
+    @property
+    def parentage(self):
+        """Yield the parent, grandparent, etc. of this Lane.
 
-        * Have an unsuppressed license pool that belongs to one of the
-          available collections.
+        The Lane may be inside one or more non-Lane WorkLists, but those
+        WorkLists are not counted in the parentage.
         """
+        seen = set()
+        if not self.parent:
+            return
+        yield self.parent
+        for parent in self.parent.parentage:
+            if parent in seen:
+                raise ValueError("Lane parentage loop detected")
+            seen.add(parent)
+            yield parent
```

Comment:
should `self` and `self.parent` also be in `seen`?

Reply:
Yes, I think so.